osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs

```diff
@@ -485,31 +485,55 @@ private void addControlPoint(double time, ControlPoint point, bool timingChange)
         private void flushPendingPoints()
         {
             // Changes from non-timing-points are added to the end of the list (see addControlPoint()) and should override any changes from timing-points (added to the start of the list).
-            // As non-timing-points should not set the OmitFirstBarline setting, we will honor the setting of the last effect point, the effect point generated by the timing-point (if any).
-            EffectControlPoint effectPoint = null;
 
             for (int i = pendingControlPoints.Count - 1; i >= 0; i--)
             {
-                if (pendingControlPoints[i] is EffectControlPoint pendingEffectPoint)
-                {
-                    if (effectPoint == null)
-                        effectPoint = pendingEffectPoint;
-                    else
-                        effectPoint.OmitFirstBarLine = pendingEffectPoint.OmitFirstBarLine;
-                }
-                else
+                var pendingPoint = pendingControlPoints[i];
+                var type = pendingPoint.GetType();
+
+                if (pendingControlPointTypes.Contains(type))
                 {
-                    var type = pendingControlPoints[i].GetType();
-                    if (pendingControlPointTypes.Contains(type))
-                        continue;
+                    // In stable, OmitFirstBarLine could only be set on timing points (ie. uninherited points).
+                    // From this perspective, it seems odd that we have the OmitFirstBarLine setting in EffectControlPoint as this breaks that assumption.
+                    //
+                    // To make sure legacy beatmaps are handled correctly, we therefore need to make sure to respect the earliest OmitFirstBarLine setting
+                    // for the current point in time (aka the one read first from the file).
+                    //
+                    // The easiest way to handle this is to transfer any non-false values from the current flush group to the actual ControlPointInfo instance.
+
+                    if ((pendingPoint as EffectControlPoint)?.OmitFirstBarLine == true)
+                    {
+                        var prevPoint = beatmap.ControlPointInfo.EffectPointAt(pendingControlPointsTime);
+
+                        // If this point is the default point, or does not share our points time, that means that the ControlPointInfo.Add() call
+                        // for the pendingPoint below did not actually add a point.
+                        //
+                        // This can happen if the point is deemed redundant due to it not changing effects.
+                        // If this happens, we need to manually re-add an effect point with OmitFirstBarLine now set to true, inheriting the rest of the settings
+                        // from the previous effect point
```

Comment:
This reads really bad. Try using the return value from `ControlPointInfo.Add` to only add to the pending list if a point was actually created.

Reply:
The problem with this is that we need the value on the next iteration. We could refrain from adding to the types list if the add failed, but then we might add a second point that normally would not have been added, breaking tests. The only way to do it this way is to get yet another variable to store the last iterations adding result, is that the way here?

Reply:
I think you have things in a pretty... best-they-can-be state. I tried a few other simplified approaches but they all have flaws in one way or another.

I do wonder if just moving this property to `TimingControlPoint` is lower effort at this point, but oh well.